dmutils/s3.py

```diff
@@ -1,74 +1,87 @@
+# import this and catch it if you like but doing so you give up any pretense at boto version independence.
+from botocore.exceptions import ClientError
 
 from .formats import DATETIME_FORMAT
 
 logger = logging.getLogger(__name__)
 
 FILE_SIZE_LIMIT = 5400000  # approximately 5Mb
 
+default_region = "eu-west-1"
+
 
 class S3(object):
-    def __init__(self, bucket_name, host='s3-eu-west-1.amazonaws.com'):
-        conn = boto.connect_s3(host=host)
+    def __init__(self, bucket_name, region_name=default_region):
+        self._resource = boto3.resource("s3", region_name=region_name)
+        self._bucket = self._resource.Bucket(bucket_name)
 
-        self.bucket_name = bucket_name
-        self.bucket = conn.get_bucket(bucket_name)
+    @property
+    def bucket_name(self):
+        return self._bucket.name
 
-    def save(self, path, file, acl='public-read', move_prefix=None, timestamp=None, download_filename=None,
+    def save(self, path, file_, acl='public-read', timestamp=None, download_filename=None,
              disposition_type='attachment'):
         """Save a file in an S3 bucket
 
         canned ACL list: https://docs.aws.amazon.com/AmazonS3/latest/dev/acl-overview.html#canned-acl
 
         :param path:              location in S3 bucket at which to save the file
-        :param file:              file object to be saved in S3
+        :param file_:             file object to be saved in S3
         :param acl:               S3 canned ACL
-        :param move_prefix:       Prefix to give to existing file when moving it out of the way
         :param timestamp:         Timestamp to set for this file rather than using utcnow
         :param download_filename: Suggested name for a browser to download, part of Content-Disposition header
         :param disposition_type:  Content-Disposition type - e.g. "attachment" or "inline"
 
         :return: S3 Key
         """
-        path = path.lstrip('/')
-
-        self._move_existing(path, move_prefix)
-
-        key = self.bucket.new_key(path)
-        filesize = get_file_size_up_to_maximum(file)
+        path = self._normalize_path(path)
         timestamp = timestamp or datetime.datetime.utcnow()
-        key.set_metadata('timestamp', timestamp.strftime(DATETIME_FORMAT))
-        headers = {'Content-Type': self._get_mimetype(key.name)}
+        filesize = get_file_size(file_)
+
+        obj = self._bucket.Object(path)
+        extra_kwargs = {}
         if download_filename:
-            headers['Content-Disposition'] = '{}; filename="{}"'.format(
-                disposition_type, download_filename
-            ).encode('utf-8')
-        key.set_contents_from_file(
-            file,
-            headers=headers
+            extra_kwargs["ContentDisposition"] = u'{}; filename="{}"'.format(
+                disposition_type,
+                # boto/aws can't cope with unicode here, but wants the ultimate result as a `str` in py3, so doing this
```

Comment:
The whole pipeline of `Content-Disposition` is a bit of a minefield from a unicode pov, right down to the user's filesystem coping with unicode... all makes me wonder if we shouldn't just `unidecode` all our suggested filenames.